Moved nupic extensions from nupic to nupic.core and updated CMakeList

akhilaananthram opened the pull request:
Moved PyRegion and the py_support helper files from nupic to nupic.core.  Updated the CMakeList file to include these files.  Also removed the need for a DynamicLibrary in RegionImplFactory to load PyRegion.cpp into nupic.core. 

There are also changes to nupic because of this.  The paths to py_support need to be updated in the SWIG files and setup.py.

nupic changes: https://github.com/numenta/nupic/pull/2313

Fixes #474

@scottpurdy

scottpurdy commented on src/CMakeLists.txt:
```diff
@@ -325,6 +335,16 @@ set(CAPNP_SPECS
+# Python support
+#
+set (PY_SUPPORT_FILES
+            nupic/py_support/NumpyVector.cpp
```
Line indentation up with previous line?

scottpurdy commented on src/CMakeLists.txt:
```diff
@@ -182,7 +185,7 @@ if(${CMAKE_CXX_COMPILER_ID} STREQUAL "MSVC")
 else()
   # clang/gcc
-  set(COMMON_COMPILE_FLAGS "-fPIC -std=c++11 -m${BITNESS} ${STDLIB} -fvisibility=hidden -Werror -Wall -Wextra -Wreturn-type -Wunused -Wno-unused-parameter -Wno-missing-field-initializers")
+  set(COMMON_COMPILE_FLAGS "-fPIC -std=c++11 -m${BITNESS} ${STDLIB} -fvisibility=hidden -Wall -Wextra -Wreturn-type -Wunused -Wno-unused-parameter -Wno-missing-field-initializers")
```
Can we keep `-Werror` in for the core C++ build and just remove it for the SWIG build?

akhilaananthram replied:
It also affects PyRegion.cpp since that uses numpy

scottpurdy commented:
Done with review.

scottpurdy commented:
Rather than `#ifdef` statements for numpy/python related code, can we just have two build rules in CMake? One, for building the dynamic python lib, would have the files that we are moving over from nupic and the other, for building C++ binaries, wouldn't.

akhilaananthram commented:
@scottpurdy: I incorporated your suggestions.  I added the -werror flag back in after updating the numpy code so it is no longer deprecated. 

Appveyor built in 39 minutes and 14 seconds before I squashed the commits.  This time though it ran out of time: "Build execution time has reached the maximum allowed time for your plan (40 minutes)."

I think it takes so long because it has to install numpy, which is why it occasionally runs out of time.  What should I do about this?

scottpurdy commented on src/CMakeLists.txt:
```diff
@@ -117,11 +117,14 @@ if("${PROJECT_BUILD_ARTIFACTS_DIR}" STREQUAL "")
 endif()
 
+find_package(PythonInterp REQUIRED)
+execute_process(COMMAND "${PYTHON_EXECUTABLE}" "-c" "import sys; import os; import numpy; sys.stdout.write(os.path.dirname(numpy.get_include()))" OUTPUT_VARIABLE NUMPY_CORE)
```
Break this into multiple lines

scottpurdy commented:
:+1: Just the one nit pick

scottpurdy commented:
@rhyolight - I have no idea how to see if this is ready to merge since I can't rerun AppVeyor

scottpurdy commented:
@akhilaananthram - can you push a dummy commit (`git commit --allow-empty` or something like that) to re-trigger AppVeyor? Since this change affects the AppVeyor config we really want to make sure we aren't breaking it!

akhilaananthram commented:
@scottpurdy: it's passing appveyor

rcrowder commented:
I've been battling with these changes this week. The file ```pyconfig.h``` in the Windows Python27 include directory, has an annoying hard-coded library import for ```Python27_d.lib``` That lib only exists if you make it yourself from source. So far I've been looking at whether Python can be disabled on MSVC Debug builds only, and/or building Python debug from source to try out that library.

rcrowder commented:
I've managed to work around this issue by rebuilding Python 2.7.9 from source using Visual Studio 2015. The resulting ```Python27_d.lib``` can then be used with NuPIC Core, and with the associated DLL moved into the right place, unit tests can be single-step debugged. Plus adds a new task to change the development notes in the Wiki and Readme.

rhyolight commented:
Sorry about the heartburn, @rcrowder. Things are getting crazy because we think we see a way forward for Windows builds. Moving SWIG to `nupic.core` is the first step. Please stay tuned and if you have any feedback, please email us at nupic-hackers so we can converse.

rcrowder commented:
From the AppVeyor logs it looks like your close to where we were just before the C++11 changes arrived. Building the x64 debug Python lib has allowed me to progress with debug builds. Just 5 failing unit tests, relating to the Python DLL not being able to import numpy. But enough to continue with TM testing.
